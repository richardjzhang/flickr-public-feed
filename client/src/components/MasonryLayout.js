// @flow strict
import styled from '@emotion/styled';
import * as React from 'react';
import { VSpacer } from 'src/components/Stack';

const FlexGrid = styled.div<{ gap: number }>((props) => ({
  display: 'flex',
  justifyContent: 'center',
  flexWrap: 'wrap',
  marginRight: -props.gap,
}));

const Col = styled.div<{ gap: number, width: number }>((props) => ({
  width: props.width,
  marginRight: props.gap,
  marginBottom: -props.gap,
}));

const constructColumns = (
  children: React$Node,
  cols: Array<Array<React$Node>>,
  gap: number,
) => {
  React.Children.forEach(children, (child, i) =>
    cols[i % cols.length].push(
      <React.Fragment key={i}>
        {child}
        <VSpacer height={gap} />
      </React.Fragment>,
    ),
  );
  return cols;
};

type Props = {|
  children: React$Node,
  gap: number,
  numberOfCols: number,
  colWidth: number,
|};

export default function MasonryLayout({
  children,
  gap,
  numberOfCols,
  colWidth,
}: Props) {
  const cols = constructColumns(
    children,
    [...Array(numberOfCols)].map(() => []),
    gap,
  );

  return (
    <FlexGrid gap={gap}>
      {[...Array(numberOfCols)].map((_, index) => (
        <Col key={index} gap={gap} width={colWidth}>
          {cols[index]}
        </Col>
      ))}
    </FlexGrid>
  );
}
