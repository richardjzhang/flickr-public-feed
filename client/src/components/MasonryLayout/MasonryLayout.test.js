import { render } from '@testing-library/react';
import * as React from 'react';

import MasonryLayout from '.';

describe('MasonryLayout', () => {
  test('renders correctly', () => {
    const tree = render(
      <MasonryLayout gap={20} numberOfCols={4} colWidth={300}>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
        <div>Test</div>
      </MasonryLayout>,
    );
    expect(tree).toMatchSnapshot();

    // There should be four columns present
    const firstColumn = tree.getByTestId('MasonryLayout-Column-0');
    const secondColumn = tree.getByTestId('MasonryLayout-Column-1');
    const thirdColumn = tree.getByTestId('MasonryLayout-Column-2');
    const fourthColumn = tree.getByTestId('MasonryLayout-Column-3');

    expect(firstColumn).toBeInTheDocument();
    expect(secondColumn).toBeInTheDocument();
    expect(thirdColumn).toBeInTheDocument();
    expect(fourthColumn).toBeInTheDocument();
  });
});
