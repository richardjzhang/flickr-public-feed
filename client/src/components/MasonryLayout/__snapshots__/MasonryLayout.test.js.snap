// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MasonryLayout renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-1a8sdri"
      >
        <div
          class="css-jnid37"
          data-testid="MasonryLayout-Column-0"
          width="300"
        >
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
        </div>
        <div
          class="css-jnid37"
          data-testid="MasonryLayout-Column-1"
          width="300"
        >
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
        </div>
        <div
          class="css-jnid37"
          data-testid="MasonryLayout-Column-2"
          width="300"
        >
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
        </div>
        <div
          class="css-jnid37"
          data-testid="MasonryLayout-Column-3"
          width="300"
        >
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
          <div>
            Test
          </div>
          <div
            class="css-6gf715-VSpacer"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-1a8sdri"
    >
      <div
        class="css-jnid37"
        data-testid="MasonryLayout-Column-0"
        width="300"
      >
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
      </div>
      <div
        class="css-jnid37"
        data-testid="MasonryLayout-Column-1"
        width="300"
      >
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
      </div>
      <div
        class="css-jnid37"
        data-testid="MasonryLayout-Column-2"
        width="300"
      >
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
      </div>
      <div
        class="css-jnid37"
        data-testid="MasonryLayout-Column-3"
        width="300"
      >
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
        <div>
          Test
        </div>
        <div
          class="css-6gf715-VSpacer"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
